herd/AArch64Sem.ml

```diff
@@ -2382,7 +2331,67 @@ module Make
         (* | I_BL _|I_BLR _|I_BR _|I_RET _ *)
         | I_LD1M _|I_ST1M _) as i ->
             Warn.fatal "illegal instruction: %s" (AArch64.dump_instruction i)
-        )
+
+(* At the moment, NOP is the onely instruction that can be used
+ * to overwrite code
+ *)
+
+      let check_self ii =
+        let inst = ii.A.inst in
+        if not (AArch64Base.is_overwritable inst) then
+          do_build_semantics inst ii
+        else match Label.norm ii.A.labels with
+        | None -> do_build_semantics inst ii
+        | Some hd ->
+           let inst_val = V.instructionToV inst in
+           (* Shadow default control sequencing operator *)
+           let(>>*=) = M.bind_control_set_data_input_first in
+           let a_v = make_label_value ii.A.fetch_proc hd in
+           let a = (* Normalised address of instruction *)
+             A.Location_global (make_label_value ii.A.fetch_proc hd) in
+             read_loc_instr a ii
+             >>= fun actual_val ->
+             M.op Op.Eq actual_val inst_val
+             >>==
+               fun cond ->
+                 M.choiceT cond
+                   (commit_pred ii
+                    >>*= fun () -> M.mk_singleton_es (Act.NoAction) ii
+                    >>= fun () ->
+                      do_build_semantics inst ii )
```

Comment:
but we call do_build_semantics, so depending on `inst` we expect `do_build_semantics` to add the right action, don't we? If `inst` is a NOP, then do_build_semantics should add an Act.NoAction.

Reply:
Yes, sounds good to me.

Reply:
@relokin I am not sure I understand: do you suggest that `do_build_semantics` adds the `No.Action` event for NOP's and Branches, bu only in `-variant self` mode? Or something else?

Reply:
Sort of. 

I had a quick chat with @artkhyzha and he clarified that we don't support overwrites where the new instruction can be anything other than a NOP (I was under the impression that we supported NOP -> B). With the current assumption this works. This action will be added only when cond = true and that means that the new instruction is a NOP. If the new instruction could be a branch this would this NoAction would be added in before of the B event.

Related, why do we handle this differently in self?  In self, I understand why execution graphs of NOPs without this extra action would look weird, but since we decided that a NOP will result in a NoAction in self, shouldn't we have it regardless of self?

Reply:
Ok , let us add this NoAction action to all instructions that generate no event at the moment. This concerns at least NOP and non-conditional branches.

Reply:
Just added NoAction as the action of NOP and branches.